text/0000-century-of-the-self-type.md

```diff
@@ -0,0 +1,178 @@
+- Feature Name: arbitrary_self_types
+- Start Date: 2018-02-15
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow types that implement `Deref` targeting `Self` to be the receiver of a
+method. If the receiver type also implements the correct `CoerceUnsized` bound,
+that method is object safe.
+
+# Motivation
+[motivation]: #motivation
+
+Today, methods can only be received by value, by reference, by mutable
+reference, or by a `Box<Self>`. This has always intended to be generalized to
+support any kind of pointer, such as an `Rc<Self>` or an `Arc<Self>`. Since
+late 2017, it has been available on nightly under the `arbitrary_self_types`
+feature.
+
+This feature is increasingly relevant because of the role of special pointer
+types to constraint self-referential types, such as generators containing
+internal references. Because different kinds of "smart pointers" can constrain
+the semantics in non trivial ways, traits can rely on certain assumptions about
+the receiver of their method, whereas just implementing the trait *for* a smart
+pointer doesn't allow that kind of reliance.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When declaring a method, users can declare the type of the `self` receiver to
+be any type `T` where `T: Deref<Target = Self>`. Shorthand exists, so that
+`self` with no ascription is of type `Self`, and `&self` and `&mut self` are of
+type `&Self` and `&mut Self`:
+
+```rust
+// All of these are valid:
+trait Foo {
+    fn by_value(self: Self);
+    fn by_ref(self: &Self);
+    fn by_ref_mut(self: &mut Self);
+    fn by_box(self: Box<Self>);
+    fn by_rc(self: Rc<Self>);
+    fn by_arc(self: Arc<Self>);
+}
+```
```

Comment:
How about HKTs in the future (assuming we get HKTs which we might not)?

Reply:
If we have a bound where `Self<A>` derefs to `Self`, I don't see why not.

Reply:
But `Self` is not of the kind of values, it has kind `* -> *`, so you shouldn't be able to construct a value of `Self`at all in this case.